compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java

```diff
@@ -322,89 +323,149 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
+                    BType queryResultType = new BArrayType(selectType);
+                    resolvedType = getResolvedType(queryResultType, type, isReadonly, env);
+                    break;
+                case TypeTags.TABLE:
+                    BType tableConstraint = types.getSafeType(((BTableType) type).constraint,
+                            true, true);
+                    selectType = checkExprSilent(selectExp, env, tableConstraint, data);
+                    if (selectType == symTable.semanticError) {
+                        errorTypes.add(tableConstraint);
+                        continue;
+                    }
+                    resolvedType = getResolvedType(symTable.tableType, type, isReadonly, env);
+                    break;
+                case TypeTags.STREAM:
+                    BType streamConstraint = types.getSafeType(((BStreamType) type).constraint,
+                            true, true);
+                    selectType = checkExprSilent(selectExp, env, streamConstraint, data);
+                    if (selectType == symTable.semanticError) {
+                        errorTypes.add(streamConstraint);
+                        continue;
+                    }
+                    resolvedType = symTable.streamType;
+                    break;
+                case TypeTags.MAP:
+                    List<BTupleMember> memberTypeList = new ArrayList<>(2);
+                    BVarSymbol stringVarSymbol = new BVarSymbol(0, null, null,
+                            symTable.semanticError, null, symTable.builtinPos, SymbolOrigin.VIRTUAL);
+                    memberTypeList.add(new BTupleMember(symTable.stringType, stringVarSymbol));
+                    BType memberType = ((BMapType) type).getConstraint();
+                    BVarSymbol varSymbol = Symbols.createVarSymbolForTupleMember(memberType);
+                    memberTypeList.add(new BTupleMember(memberType, varSymbol));
+                    BTupleType newExpType = new BTupleType(null, memberTypeList);
+                    selectType = checkExprSilent(selectExp, env, newExpType, data);
+                    if (selectType == symTable.semanticError) {
+                        errorTypes.add(newExpType);
+                        continue;
+                    }
+                    resolvedType = getResolvedType(selectType, type, isReadonly, env);
+                    break;
+                case TypeTags.STRING:
+                case TypeTags.XML:
+                case TypeTags.XML_COMMENT:
+                case TypeTags.XML_ELEMENT:
+                case TypeTags.XML_PI:
+                case TypeTags.XML_TEXT:
+                    selectType = checkExprSilent(selectExp, env, type, data);
+                    if (selectType == symTable.semanticError) {
+                        errorTypes.add(type);
+                        continue;
+                    }
+
+                    if (Types.getReferredType(selectType).getKind() == type.getKind()
```

Comment:
Can't we use tag here also?

Reply:
no.

Reply:
In here we check whether the `selectType` is a subtype of the` type`. To do that we need to have compare it using kinds.

Reply:
And you can use type tag for that? We usually use type tag instead of kind.

Reply:
In here we check whether the selectType is a subtype of the type. To do that we need to have compare it using kinds.

Reply:
Is it for subtypes of string and xml? I would use org.wso2.ballerinalang.compiler.util.TypeTags#isXMLTypeTag, org.wso2.ballerinalang.compiler.util.TypeTags#isStringTypeTag.

Reply:
`ambiguous type '(Baz|Qux)'` will be the only error.

Reply:
Yes, this is the correct approach. Thanks for pointing. Updated in [#commit](https://github.com/ballerina-platform/ballerina-lang/pull/40835/commits/c16b3589fdb2c797c993858921bef3f08f49a3ba).